### ISSUE TITLE
Windows 7 installation failed 

### ISSUE THREAD

**Issue by sergsb** [hughperkins/DeepCL#25]
I've faced with an error during installation on Windows 7, python 2.7(from Anaconda). 
pip install deepcl or python setup.py install resulting...
![deepcl_error](https://cloud.githubusercontent.com/assets/4963384/8283265/fb34c98a-18fe-11e5-959b-d4ffef8cc0c2.png)
Is it a bug in installation script?

**Comment by sergsb** [comment]
Proved in Linux too.

**Comment by hughperkins** [comment]
Yes, ok.  I need to fix this.  Trying to implement char-rnn in https://github.com/hughperkins/clnn currently.  I havent forgotten you :-)

**Comment by hughperkins** [comment]
Hmmm, ok, exists on linux too, and I can replicate it (just have).  That makes it easier to fix.

**Comment by hughperkins** [comment]
Hmmm, what happens if you download the egg directly from https://pypi.python.org/pypi/DeepCL/5.10.2 , and then do `easy_install DeepCL-5.10.2-py2.7-linux-x86_64.egg`? (or equivalent, depending on os/bitness).

**Comment by hughperkins** [comment]
![deepcl-py-installokfromegg](https://cloud.githubusercontent.com/assets/123560/8344079/f2049f68-1b0e-11e5-8e67-1fe5190cf70e.png)

**Comment by hughperkins** [comment]
(Fixed the setup.py so it creates a valid source distribution; I need to rerun the builds to upload this to pypi and so on)

**Comment by hughperkins** [comment]
Got the whole thing building on windows and linux again, and done a new release.  For some unknown reason, pip is still downloading the tarball, rather than the egg. I'm not sure why?  Anyway, it seems it builds ok now?

**Comment by rilut** [comment]
I have this problem on installing DeepCL from pip. From what I heard, /EHsc is a specific parameter for VS C++. 

```
{lamb} pip install deepcl --pre --upgrade
Collecting deepcl
  Using cached DeepCL-8.0.0.tar.gz
Building wheels for collected packages: deepcl
  Running setup.py bdist_wheel for deepcl
  Complete output from command C:\Anaconda3\python.exe -c "import setuptools;__file__='C:\\Users\\rilut\\AppData\\Local\\Temp\\pip-build-504kqvm_\\deepcl\\setup.py';exec(compile(open(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" bdist_wheel -d C:\Users\rilut\AppData\Local\Temp\tmp6tago5d3pip-wheel-:
  version:  8.0.0
  running bdist_wheel
  running build
  running build_ext
  building 'PyDeepCL' extension
  creating build
  creating build\temp.win-amd64-3.4
  creating build\temp.win-amd64-3.4\Release
  C:\Anaconda3\pkgs\mingw-4.7-1\MinGW\bin\gcc.exe -mdll -O -Wall -Imysrc -Imysrc/lua -IC:\Anaconda3\include -IC:\Anaconda3\include -c PyDeepCL.cxx -o build\temp.win-amd64-3.4\Release\pydeepcl.o /EHsc
  gcc: error: /EHsc: No such file or directory
  error: command 'C:\\Anaconda3\\pkgs\\mingw-4.7-1\\MinGW\\bin\\gcc.exe' failed with exit status 1

  ----------------------------------------
  Failed building wheel for deepcl
Failed to build deepcl
```

How to make pip compile DeepCL with VS C++  rather than gcc? I've installed VS 2013 Express.

**Comment by hughperkins** [comment]
seems like you are running pip from mingw?  So, its using mingw gcc to compile.

I think you'd need to install python from cmd, and use python from a cmd, for it to be built using visual studio.

By the way, it wont build with vs 2013, as far as I know:
- Python 2.7 => builds using visual studio 2008
- Python 3.4 => builds using visual studio 2010

You need the exact same version of visaul studio, AFAIK.

**Comment by rilut** [comment]
@hughperkins no, it's from `cmder`. I just tested it with `cmd`, but it's the same.
Should DeepCL be building with `cl` instead of `gcc` in windows? Why it was looking for `gcc`?
Did my PATH is polluted or what?

**Comment by hughperkins** [comment]
oh wait... Windows ... DeepCL... so the way this works is:
- open cmake.  its a gui client.  from cmake open the cmakelists.txt and so on, do the configure and generate.  it'll ask you what generator you want to use.  You'll specifci visual studio, and choose a verson (eg 2013 64-bit)
- open visual studio, and open the project/solution files ,and build them.

**Comment by rilut** [comment]
@hughperkins Wow thanks for the detailed comment, I'll look into it.

I'm using Anaconda 2.2.0 (64-bit) (Python 3.4.3). But I think I also have a lot of things in my PATH

**Comment by hughperkins** [comment]
Hi rilut,

Quick update: just noticed that the jenkins build was actually broken :-P  I've updated it just now, and tested it on win32 py2.7, win32 py3.4, and win64 py2.7 (win64 py3.4 still running as I write), and it passes again now.  The new jenkins scripts look like (in the [jenkins](https://github.com/hughperkins/DeepCL/tree/master/jenkins) directory):

eg [win64-py27.bat](https://github.com/hughperkins/DeepCL/blob/master/jenkins/win64-py27.bat)
```

call %~dp0win-py.bat 64 27
if errorlevel 1 goto :fail
goto :eof

:fail
echo Something went wrong
exit /B 1
```

[win-py.bat](https://github.com/hughperkins/DeepCL/blob/master/jenkins/win-py.bat)
```
echo args: %1 %2
set bitness=%1
set pyversion=%2
echo bitness: %bitness%
echo pyversion: %pyversion%

call %~dp0win-cpp.bat %bitness%

call \env-%pyversion%-%bitness%\scripts\activate

python -c "from __future__ import print_function; import platform; print( platform.uname() )"
python -c "from __future__ import print_function; import platform; print( platform.architecture() )"

rem  cd
rem  rmdir /s /q build
rem  rmdir /s /q dist
rem  dir
rem  mkdir build
rem  cd build
rem  set "generatorpostfix="
rem  if %bitness%==64 set "generatorpostfix= Win64"
rem  "c:\program files (x86)\cmake\bin\cmake" -G "Visual Studio 10 2010%generatorpostfix%" ..
rem  C:\WINDOWS\Microsoft.NET\Framework\v4.0.30319\MSBuild.exe ALL_BUILD.vcxproj /p:Configuration=Release
rem  if errorlevel 1 exit /B 1
rem  C:\WINDOWS\Microsoft.NET\Framework\v4.0.30319\MSBuild.exe INSTALL.vcxproj /p:Configuration=Release
rem  if errorlevel 1 exit /B 1
rem  cd ..
rem  cd
rem  dir
cd %~dp0..
call dist\bin\activate.bat

copy /y jenkins\version.txt python
cd python

rmdir /s /q dist
rmdir /s /q build
rmdir /s /q mysrc
rmdir /s /q src
dir
if exist dist goto :error
if exist build goto :error
if exist mysrc goto :error
if exist src goto :error

set
python setup.py build_ext -i
if errorlevel 1 goto :error

python setup.py bdist_egg
if errorlevel 1 goto :error

set HOME=%HOMEPATH%
python setup.py bdist_egg upload
rem ignore any error?
rem if errorlevel 1 goto :error

exit /B 0

goto :eof

:error
echo Error occurred
exit /B 1
```

[win-cpp.bat](https://github.com/hughperkins/DeepCL/blob/master/jenkins/win-cpp.bat)
```
set WINBITS=%1
echo WINBITS: %WINBITS%

cd %~dp0.
for /f "" %%i in (version.txt) do (
   set version=%%i
)

cd %~dp0..
if not exist turbogjpeg-win%WINBITS%.zip powershell.exe -Command (new-object System.Net.WebClient).DownloadFile('http://deepcl.hughperkins.com/Downloads/turbojpeg-1.4.0-win%WINBITS%-static.zip', 'turbojpeg-win%WINBITS%.zip')
if errorlevel 1 exit /B 1
rmdir /s /q turbojpeg-win%WINBITS%
mkdir turbojpeg-win%WINBITS%
cd turbojpeg-win%WINBITS%
"c:\program files\7-Zip\7z.exe" x ..\turbojpeg-win%WINBITS%.zip
if errorlevel 1 exit /B 1

cd %~dp0..
dir
rmdir /s /q build
mkdir build
cd %~dp0..\build
dir
set "VS100COMNTOOLS=c:\Program Files (x86)\Microsoft Visual Studio 10.0\Common7\Tools\"
set "VS110COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio 11.0\Common7\Tools\"
set "VS120COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio 12.0\Common7\Tools\"
echo get_filename_component(SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR} DIRECTORY)>initcache.cmake
echo set(JPEG_INCLUDE_DIR "${SOURCE_DIR}/turbojpeg-win%WINBITS%" CACHE PATH "JPEG_INCLUDE_DIR")>>initcache.cmake
echo set(JPEG_LIBRARY "${SOURCE_DIR}/turbojpeg-win%WINBITS%/turbojpeg-static.lib" CACHE PATH "JPEG_LIBRARY")>>initcache.cmake
if exist "c:\program files\cmake\bin\cmake.exe" set "CMAKEEXE=c:\program files\cmake\bin\cmake.exe"
if exist "c:\program files (x86)\cmake\bin\cmake.exe" set "CMAKEEXE=c:\program files (x86)\cmake\bin\cmake.exe"
set "generatorpostfix="
if x%WINBITS%==x64 set "generatorpostfix= Win64"
"%CMAKEEXE%" -G "Visual Studio 10 2010%generatorpostfix%" -C initcache.cmake ..
C:\WINDOWS\Microsoft.NET\Framework\v4.0.30319\MSBuild.exe ALL_BUILD.vcxproj /p:Configuration=Release
if errorlevel 1 exit /B 1
C:\WINDOWS\Microsoft.NET\Framework\v4.0.30319\MSBuild.exe INSTALL.vcxproj /p:Configuration=Release
if errorlevel 1 exit /B 1

rem copy down the redistributables (maybe they're on the server somewhere?)
cd %~dp0..
powershell Set-ExecutionPolicy unrestricted
if not exist vc2010redist.zip powershell.exe -Command (new-object System.Net.WebClient).DownloadFile('http://deepcl.hughperkins.com/Downloads/vc2010redist.zip', 'vc2010redist.zip')
if errorlevel 1 exit /B 1

rmdir /s /q vc2010redist
"c:\program files\7-Zip\7z.exe" x vc2010redist.zip
if errorlevel 1 exit /B 1

copy vc2010redist\win%WINBITS%\* dist\bin

cd %~dp0..
"c:\program files\7-Zip\7z.exe" a deepcl-win%WINBITS%-%version%.zip dist
if errorlevel 1 exit /B 1

cd %~dp0..
echo %version%>latestUnstable.txt
```

tldr; :
- first use ccmake to create the visual studio project files
- then, open visual studio, and build the visual studio project files
- then, open a command prompt, and do:
  - activate your python virtualenvironment (if appropriate)
  - call the `dist\activate.bat`, from the build output just now
  - cd into the `python` directory, and do `python setup.py install`

**Comment by hughperkins** [comment]
Hi, I've deleted all my useless replies, and updated the README.md for Python, iwth instructions for building from source :-) https://github.com/hughperkins/DeepCL/blob/master/python/README.md

Basically:
- first build the native libraries, see https://github.com/hughperkins/DeepCL/blob/master/doc/Build.md
- then open a cmd:
  - call `dist\bin\activate.bat`
  - cd into `python` directory
  - run `python setup.py install`
=> hopefully should work.... if still not working, then get back to me, and I'll work through it a bit.

(PS, in case useful, as before, here is a link to the jenkins scripts, as before, but which I fixed just now, and tested, and they are working now, eg https://github.com/hughperkins/DeepCL/blob/master/jenkins/win64-py34.bat (which calls some other scripts from the same directory) They were broken earlier actually...)

**Comment by hughperkins** [comment]
> I have this problem on installing DeepCL from pip.

Hmmm.... in theory.... installing from pip... should just be an install, no build.  Perhaps you can try directly downloading the egg, and installing that?  egg is eg: https://pypi.python.org/packages/3.4/D/DeepCL/DeepCL-8.1.2-py3.4-win-amd64.egg#md5=a6fe39f92d454abb32a2f74c5dcd9c3f (from http://deepcl.hughperkins.com/Downloads/ )

(PS sorry about the disorganized replies; I'm having to remember how the python builds work :-D )
